designs/0033-gradients-framework.md

```diff
@@ -0,0 +1,232 @@
+
+```cpp
+  reverse_pass_callback(
+      [rev_grad_fun_tuple, arena_tuple, prim_tuple, rtn]() mutable {
+        // Iterate over input arguments, applying the respective gradient
+        // function with the tuple of extracted primitive values
+        walk_tuples(
+            [&](auto&& grad_funs, auto&& arg) {
+              // Only calculate gradients if the input argument is not primitive
+              if (!is_constant_all<decltype(arg)>::value) {
+                // Need to wrap the argument in a forward_as<var>() so that it
+                // will compile with both primitive and var inputs
+                forward_as<promote_scalar_t<var, decltype(arg)>>(arg).adj() +=
+                    // Use the relevant gradient function with the tuple of
+                    // primitive arguments
+                    math::apply(
+                        [&](auto&&... args) {
+                          return grad_funs(rtn.val_op(), rtn.adj_op(),
+                                   internal::arena_val(
+                                       std::forward<decltype(args)>(args))...);
+                        },
+                        prim_tuple);
+              }
+            },
+            std::forward<RevGradFunT>(rev_grad_fun_tuple),
+            std::forward<decltype(arena_tuple)>(arena_tuple));
+      });
+```
+
+Where `walk_tuples` is a functor for the parallel iteration over multiple tuples (parallel as in `pmap`, rather than multi-threaded).
+
+## Forward-Mode
+
+A similar approach is used for forward-mode, but instead the gradients are accumulated into a single variable which is then used to construct the return `fvar`:
+
+```cpp
+  auto d_ = internal::initialize_grad(std::forward<rtn_t>(rtn));
+
+  walk_tuples(
+      [&](auto&& f, auto&& arg, auto&& dummy) {
+        using arg_t = decltype(arg);
+        if (!is_constant_all<arg_t>::value) {
+          d_ += math::apply(
+              [&](auto&&... args) {
+                return f(rtn,
+                         forward_as<promote_scalar_t<ReturnT, arg_t>>(arg).d(),
+                         args...);
+              },
+              val_tuple);
+        }
+      },
+      std::forward<FwdGradFunT>(fwd_grad_fun_tuple),
+      std::forward<ArgsTupleT>(args_tuple),
+      std::forward<decltype(dummy_tuple)>(dummy_tuple));
+
+  return to_fvar(rtn, d_);
+```
+
+Where `internal::initialize_grad` is used to initialise a variable with the same type and dimension as the return value.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This does appear to increase compilation time - given the increased complexity of templating and tuple manipulation
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The best alternative is the current approach, of separate specifications for `prim`, `fwd`, and `rev`. But this has the limitations mentioned in the Motivation section.
+
+# Prior art
+[prior-art]: #prior-art
+
+None that I'm aware of
```

Comment:
This _kind of_ reminds me of some of the things in `FastAD`

https://github.com/JamesYang007/FastAD